### PR TITLE
Add server error handler

### PR DESCRIPTION
**Pull request by zoltrain**
NOTE: All development is currently focused on Sapper's successor, SvelteKit: https://kit.svelte.dev/
We are not adding new features to Sapper. While we might fix some particularly pressing or easy to
address bugs, it's highly unlikely that most PRs will be reviewed.

### Before submitting the PR, please make sure you do the following
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ ] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
-  [ ] Run the tests with `npm test` and lint the project with `npm run lint`

### CODE REVIEW

File: runtime/src/server/middleware/get_server_route_handler.ts [hunk 1]
```diff
@@ -1,76 +1,91 @@
...
-					res.end(err.message);
-				} else {
-					process.nextTick(next);
-				}
-			};
+      const handle_next = (err?: Error) => {
+        if (err) {
+          res.statusCode = 500;
+          res.end(err.message);
+          const ne = next as { (err?: Error | string): undefined | Error };
```
Comment (zoltrain) [inline runtime/src/server/middleware/get_server_route_handler.ts]: actually call next with an error (express style).

File: runtime/src/server/middleware/index.ts [hunk 2]
```diff
@@ -1,143 +1,235 @@
...
-		serve({
-			prefix: '/client/',
-			cache_control: dev ? 'no-cache' : 'max-age=31536000, immutable'
-		}),
-
-		get_server_route_handler(manifest.server_routes),
-
-		get_page_handler(manifest, session || noop)
-	].filter(Boolean));
+export function middleware(
```
Comment (zoltrain) [inline runtime/src/server/middleware/index.ts]: keep the status quo

File: runtime/src/server/middleware/index.ts [hunk 3]
```diff
@@ -1,143 +1,235 @@
...
-	return !ignore
-		? (req, res, next) => nth_handler(0, req, res, next)
-		: (req, res, next) => {
-			if (should_ignore(req.path, ignore)) {
-				next();
-			} else {
-				nth_handler(0, req, res, next);
-			}
-		};
+export function middlewares(
```
Comment (zoltrain) [inline runtime/src/server/middleware/index.ts]: only return middleware functions

File: runtime/src/server/middleware/index.ts [hunk 4]
```diff
@@ -1,143 +1,235 @@
...
+
+  function nth_handler(
+    n: number,
+    req: SapperRequest,
+    res: SapperResponse,
+    next: () => void
+  ) {
+    const ne = next as { (err?: Error | string): Error | undefined };
+    if (n >= total) {
+      return ne();
```
Comment (zoltrain) [inline runtime/src/server/middleware/index.ts]: This is the source of all the error misery, errors are not propagated to any middleware after the last sapper middleware exits

**Review by seanlail — APPROVED** [review]
🎉

**Pull request merged**